portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -30,11 +33,18 @@
 // TODO: change class name, abstract comment to a class on its own and separate 'post' and 'get' servlets
 @WebServlet("/comment")
 public class DataServlet extends HttpServlet {
-  private final List<String> comments = new ArrayList<>();
+  private final Query query = new Query("Comment").addSort("timestamp", SortDirection.ASCENDING);
```

Comment:
We shouldn't be storing the query as a class `final` field, as we this query is not something that's quite globally applicable, and we may have quite a few queries in one servlet.

Reply:
Oh yes. This is true. Thanks for pointing it out! (:

Reply:
The point about making something a final field for the class is that we want to avoid its initialization overhead, and mostly it's some higher level util/helper/manager/controller classes which are applicable or used by a few methods, like the Gson and DataStoreService. For a query, first it's only applicable to this the `doGet` method. Second that this query is configurable. For example, let's say one day we want to return a list of comments sorted by timestamp in descending order, how are we going to store the queries?

Reply:
I see! Thanks for the clarification. :D I do agree with what you have said. I will change it now. ☺️

Reply:
Hmmm I don't mean just removing the `final` field, but we should move this into `doGet` method, as there isn't any point storing it as class field now.

Reply:
Also, the DataStoreService should be moved to class final field, similar to `Gson`

Reply:
Oh I see. However, in this case, the get request will always be performing the same query wouldn't it? 

Oh yes, nice catch! I will move that in as well (:

Reply:
I have made the changes you have mentioned but it would be great if you could still clarify your first comment. Thanks in advance! :D

Reply:
Sorry can you clarify a bit on what you mean by my first comment?

Reply:
Oh. I am referring to "Hmmm I don't mean just removing the final field, but we should move this into doGet method, as there isn't any point storing it as class field now."

Reply:
I see you already made the change in https://github.com/pohlinwei/software-product-sprint/pull/14/commits/4a4cc3fa685e02cb9e4242a28b76b284755bba60, so do you mean that you don't quite understand why this change is needed?

Reply:
Yup!

Reply:
So to clarify things here, we are talking about why we need shouldn't make `Query` a class level field.

I've explained in earlier comment
> The point about making something a final field for the class is that we want to avoid its initialization overhead, and mostly it's some higher level util/helper/manager/controller classes which are applicable or used by a few methods, like the Gson and DataStoreService. For a query, first it's only applicable to this the doGet method. Second that this query is configurable. For example, let's say one day we want to return a list of comments sorted by timestamp in descending order, how are we going to store the queries?

I guess you are confused by the differences between the following three:
```
class DataServlet extends HttpServlet {
   private final Query = new Query("...")...;
   
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
        PreparedQuery results = datastore.prepare(query);
        ...
   }
}
```

```
class DataServlet extends HttpServlet {
   private Query = new Query("...")...;
   
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
        PreparedQuery results = datastore.prepare(query);
        ...
   }
}
```

```
class DataServlet extends HttpServlet {
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
        Query query = new Query(...)...;
        PreparedQuery results = datastore.prepare(query);
        ...
   }
}
```
The difference between the first two is just the `final` modifier, however, that simply indicates whether we expect the query to be changed, but it doesn't change the fact that this `query` is still a class level field and globally accessible within the class. Other methods can easily mutate the query without any restrictions.

Moving the query into `doGet` doesn't mean that the query is always the same. For example, if you want to create a different query based on URL paramteres, you can simply do some conditional queries. E.g,
```
class DataServlet extends HttpServlet {
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
        Query query;
        if (someparame == "asc") {
            query = new Query(...)...;
        } else if (someparam == "dsc") {
            query = new Query(...)...;
        } else {
            response.writeln(someerror);
            return;
        }
        PreparedQuery results = datastore.prepare(query);
        ...
       response.writeln(someerror);
   }
}
```

What you want to do with the query within method will be totally up to you.

Reply:
Ahh.. Okay. Once again, thanks!! :D